dependency_test.go

```diff
@@ -0,0 +1,260 @@
+package testcontainers
+
+import (
+	"context"
+	"fmt"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go/wait"
+	"strings"
+	"testing"
+	"time"
+)
+
+func Test_ContainerDependency(t *testing.T) {
+	type TestCase struct {
+		name                string
+		configureDependants func(ctx context.Context, t *testing.T) []ContainerDependency
+		containerRequest    ContainerRequest
+		expectedEnv         []string
+		expectedError       string
+	}
+	testCases := []TestCase{
```

Comment:
suggestion: There's lots of complexity here in the actual table and more in the main flow because of the conditional behaviour which actually results in more code than splitting into dedicated sub sets with `t.Run(...)`

We should always use succinct test names, avoiding spaces as they are replaced on the output making failures harder to track down.

Here's an example from the last item in the table.
```go
	t.Run("empty-key", func(t *testing.T) {
			nginxReq := ContainerRequest{
				Image:        nginxAlpineImage,
				ExposedPorts: []string{nginxDefaultPort},
				WaitingFor:   wait.ForListeningPort(nginxDefaultPort),
			}
			dependency := NewContainerDependency(nginxReq, "")
			c, err := GenericContainer(ctx, GenericContainerRequest{
				ContainerRequest: tc.containerRequest,
				Started:          true,
				DependsOn:  []ContainerDependency{dependency},
			})
			require.EqualError(t, err, "cannot create dependency with empty environment key")
        })
````

If there is a block of the validation logic that's repeated extract that to a helper which continues to keep things easy to follow, reducing duplication.